File: leetcode/cpp/20.valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        vector<char> stack;
        for (char c: s) {
            switch (c) {
                case ')':
                    if (stack.empty() || stack.back() != '(') {
                        return false;
                    }
                    stack.pop_back();
                    break;
                case ']':
                    if (stack.empty() || stack.back() != '[') {
                        return false;
                    }
                    stack.pop_back();
                    break;
                case '}':
                    if (stack.empty() || stack.back() != '{') {
                        return false;
                    }
                    stack.pop_back();
                    break;
                default:
                    stack.push_back(c);
                    break;
            }
        }
        return stack.size() == 0;
    }
};

/* */

class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;

        for (char& c : s) {
            if (c == '{' || c == '(' || c == '[') {
                stk.push(c);
            } else if (c == '}' || c == ']' || c == ')') {
                if (stk.empty()) return false;
                char open = stk.top();
                stk.pop();
                if ((c == '}' && open != '{') || (c == ']' && open != '[') || (c == ')' && open != '(')) return false;
            }
        }

        return stk.empty();
    }
};

File: leetcode/go/20.valid-parentheses.go
func isValid(s string) bool {
    stack := make([]rune, 0, len(s))
    for _, r := range s {
        switch r {
            case '(', '[', '{':
                stack = append(stack, r)
            case ')':
                if len(stack) == 0 || stack[len(stack)-1] != '(' {
                    return false
                }
                stack = stack[0:len(stack)-1]
            case ']':
                if len(stack) == 0 || stack[len(stack)-1] != '[' {
                    return false
                }
                stack = stack[0:len(stack)-1]
            case '}':
                if len(stack) == 0 || stack[len(stack)-1] != '{' {
                    return false
                }
                stack = stack[0:len(stack)-1]
        }
    }
    return len(stack) == 0
}

